sklearn/utils/estimator_checks.py

```diff
@@ -2089,8 +2091,25 @@ def check_estimators_data_not_an_array(name, estimator_orig, X, y):
     set_random_state(estimator_1)
     set_random_state(estimator_2)
 
-    y_ = NotAnArray(np.asarray(y))
-    X_ = NotAnArray(np.asarray(X))
+    if obj_type not in ["NotAnArray", 'PandasDataframe']:
+        raise ValueError("Data type {0} not supported".format(obj_type))
+
+    if obj_type == "NotAnArray":
+        y_ = NotAnArray(np.asarray(y))
+        X_ = NotAnArray(np.asarray(X))
+    else:
+        try:
```

Comment:
I think add a comment that we explicitly test a pandas DataFrame since some estimators may handle them (especially their indexing) specially.